File: php/features.php
<?php
declare(strict_types=1);

// MixpanelAnnotations SDK feature factory

require_once __DIR__ . '/feature/BaseFeature.php';
require_once __DIR__ . '/feature/DebugFeature.php';
require_once __DIR__ . '/feature/IdempotencyFeature.php';
require_once __DIR__ . '/feature/MetricsFeature.php';
require_once __DIR__ . '/feature/PagingFeature.php';
require_once __DIR__ . '/feature/RatelimitFeature.php';
require_once __DIR__ . '/feature/RetryFeature.php';
require_once __DIR__ . '/feature/TestFeature.php';
require_once __DIR__ . '/feature/TimeoutFeature.php';


class MixpanelAnnotationsFeatures
{
    public static function make_feature(string $name)
    {
        switch ($name) {
            case "base":
                return new MixpanelAnnotationsBaseFeature();
            case "debug":
                return new MixpanelAnnotationsDebugFeature();
            case "idempotency":
                return new MixpanelAnnotationsIdempotencyFeature();
            case "metrics":
                return new MixpanelAnnotationsMetricsFeature();
            case "paging":
                return new MixpanelAnnotationsPagingFeature();
            case "ratelimit":
                return new MixpanelAnnotationsRatelimitFeature();
            case "retry":
                return new MixpanelAnnotationsRetryFeature();
            case "test":
                return new MixpanelAnnotationsTestFeature();
            case "timeout":
                return new MixpanelAnnotationsTimeoutFeature();
            default:
                return new MixpanelAnnotationsBaseFeature();
        }
    }

    /**
     * Does a generated feature class back this name? False for a name only
     * an options extend instance can supply (the station adopt path) - the
     * constructor uses this to skip make_feature for such names instead of
     * adding a stray BaseFeature.
     */
    public static function has_feature(string $name): bool
    {
        switch ($name) {
            case "base":
            case "debug":
            case "idempotency":
            case "metrics":
            case "paging":
            case "ratelimit":
            case "retry":
            case "test":
            case "timeout":
                return true;
            default:
                return false;
        }
    }
}

File: py/mixpanelannotations_sdk/features.py
# MixpanelAnnotations SDK feature factory

from mixpanelannotations_sdk.feature.base_feature import MixpanelAnnotationsBaseFeature
from mixpanelannotations_sdk.feature.debug_feature import MixpanelAnnotationsDebugFeature
from mixpanelannotations_sdk.feature.idempotency_feature import MixpanelAnnotationsIdempotencyFeature
from mixpanelannotations_sdk.feature.metrics_feature import MixpanelAnnotationsMetricsFeature
from mixpanelannotations_sdk.feature.paging_feature import MixpanelAnnotationsPagingFeature
from mixpanelannotations_sdk.feature.ratelimit_feature import MixpanelAnnotationsRatelimitFeature
from mixpanelannotations_sdk.feature.retry_feature import MixpanelAnnotationsRetryFeature
from mixpanelannotations_sdk.feature.test_feature import MixpanelAnnotationsTestFeature
from mixpanelannotations_sdk.feature.timeout_feature import MixpanelAnnotationsTimeoutFeature


_FEATURES = {
    "base": lambda: MixpanelAnnotationsBaseFeature(),
    "debug": lambda: MixpanelAnnotationsDebugFeature(),
    "idempotency": lambda: MixpanelAnnotationsIdempotencyFeature(),
    "metrics": lambda: MixpanelAnnotationsMetricsFeature(),
    "paging": lambda: MixpanelAnnotationsPagingFeature(),
    "ratelimit": lambda: MixpanelAnnotationsRatelimitFeature(),
    "retry": lambda: MixpanelAnnotationsRetryFeature(),
    "test": lambda: MixpanelAnnotationsTestFeature(),
    "timeout": lambda: MixpanelAnnotationsTimeoutFeature(),
}


def _make_feature(name):
    factory = _FEATURES.get(name)
    if factory is not None:
        return factory()
    return _FEATURES["base"]()


# True when this SDK was generated with the named feature class - the
# constructor's tolerance for extend-carried features reads this (an
# active name with no generated class must not become a BaseFeature
# stray when an extend instance carries it).
def _has_feature(name):
    return name in _FEATURES
